### PR TITLE
Add support for arbitrary metadata with the cp command

### PR DESCRIPTION
**Pull request by JordonPhillips**
This will allow a `--metadata` flag for `aws s3 cp` with the same interface as `aws s3api put-object --metadata`.

Example Usage:
```sh
$ aws s3 cp ~/test.txt s3://test-bucket/ --metadata some=value
upload: test.txt to s3://test-bucket/test.txt
$ aws s3api head-object --bucket test-bucket --key test.txt
{
    "AcceptRanges": "bytes",
    "ContentType": "text/plain",
    "LastModified": "Fri, 20 Nov 2015 21:13:12 GMT",
    "ContentLength": 13,
    "ETag": "\"ETAG\"",
    "Metadata": {
        "some": "value"
    }
}
$ aws s3 cp ~/test2.txt s3://test-bucket/ --metadata '{"other":"value"}'
upload: test2.txt to s3://test-bucket/test2.txt
$ aws s3api head-object --bucket test-bucket --key test2.txt
{
    "AcceptRanges": "bytes",
    "ContentType": "text/plain",
    "LastModified": "Fri, 20 Nov 2015 21:15:04 GMT",
    "ContentLength": 13,
    "ETag": "\"ETAG\"",
    "Metadata": {
        "other": "value"
    }
}
```

### CODE REVIEW

**Comment by jamesls** [comment]
What's the expected result for an s3->s3 copy?  Should we require that the user also has to provide the appropriate value for `--metadata-directive`?

**Comment by JordonPhillips** [comment]
@jamesls Yes, it requires that `--metadata-directive` be set. I mention that in the doc string.

**Comment by jamesls** [comment]
Yeah, that's more what I was getting at.  Would it ever be valid for an s3->s3 transfer where the user provides `--metadata` but not `--metadata-directive`?  If this is something that should always be required for s3->s3 sync, are there any downsides to setting metadata-directive=REPLACE for the user when we detect this case?

**Comment by kyleknap** [comment]
I would be in favor of automatically setting it to ``REPLACE`` and making sure users do not have to worry about it, given there is no downsides.

**Comment by JordonPhillips** [comment]
No, it would never be valid to use without `--metadata-directive`. I'll go ahead and default it.

File: awscli/customizations/s3/subcommands.py [hunk 1]
```diff
@@ -649,7 +665,7 @@ class CpCommand(S3TransferCommand):
             "or <S3Uri> <S3Uri>"
     ARG_TABLE = [{'name': 'paths', 'nargs': 2, 'positional_arg': True,
                   'synopsis': USAGE}] + TRANSFER_ARGS + \
-                [METADATA_DIRECTIVE, EXPECTED_SIZE, RECURSIVE]
+                [METADATA, METADATA_DIRECTIVE, EXPECTED_SIZE, RECURSIVE]
```
Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: We should add ``METADATA`` to ``mv`` as well. @jamesls what is your thoughts on adding to ``sync`` as well? I think I would be in favor of adding it if someone had custom metadata that applied to all objects being synced.

File: awscli/customizations/s3/utils.py [hunk 2]
```diff
@@ -564,6 +565,7 @@ def map_copy_object_params(cls, request_params, cli_params):
         """Map CLI params to CopyObject request params"""
         cls._set_general_object_params(request_params, cli_params)
         cls._set_metadata_directive_param(request_params, cli_params)
+        cls._set_metadata_params(request_params, cli_params)
```
Comment (kyleknap) [inline awscli/customizations/s3/utils.py]: We need to add it for multipart-uploads and copies as well. You can do this by setting it in ``map_create_multipart_upload``. It would be great to get an integration test as well to make sure multipart's are handled correctly.

File: awscli/customizations/s3/subcommands.py [hunk 3]
```diff
@@ -298,6 +298,22 @@
...
+    'schema': {
+        'type': 'map',
+        'key': {'type': 'string'},
+        'value': {'type': 'string'}
+    },
+    'help_text': (
+        "A map of metadata to store with the objects in S3. This will be "
+        "applied to every object which is part of this request. When copying "
+        "between two s3 locations, the metadata-directive argument will need to"
+        "be set to REPLACE."
```
Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: Update the docstring to remove the tidbit about needing to provide REPLACE for metadata-directive as we now autopopulate it.

**Comment by kyleknap** [comment]
Looks good. Had a couple comments about the handling of multipart transfers and whether we should add it to the sync command.

**Comment by JordonPhillips** [comment]
Adding it to sync would add another layer of complexity because we would have to check for metadata differences when deciding to sync. I'm not -1, but I think it would be enough work to need to be timeboxed.

File: awscli/customizations/s3/subcommands.py (hunk 1 above)
Comment (jamesls) [inline awscli/customizations/s3/subcommands.py]: By supporting it in sync, we're simply saying that we'll set metadata to whatever you've specified if we need to transfer a file to S3 right?  Seems like it would make sense.

Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: Yeah that was what I was thinking. We still do not support syncing metadata.

Comment (JordonPhillips) [inline awscli/customizations/s3/subcommands.py]: That's a much simpler goal, but we will need to explicitly say that in the documentation. An event handler would be the correct way to make this documentation modification, correct?

Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: No you would just need to update the ``help_text`` for the ``MEATADATA`` argument.

Comment (JordonPhillips) [inline awscli/customizations/s3/subcommands.py]: Do you think what I have in now looks fine?

Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: If we add it to sync, it would be good to add a note for the behavior with the sync command.

**Comment by JordonPhillips** [comment]
Docs are all updated

File: awscli/customizations/s3/subcommands.py [hunk 4]
```diff
@@ -680,7 +697,7 @@ class SyncCommand(S3TransferCommand):
     USAGE = "<LocalPath> <S3Uri> or <S3Uri> " \
             "<LocalPath> or <S3Uri> <S3Uri>"
     ARG_TABLE = [{'name': 'paths', 'nargs': 2, 'positional_arg': True,
-                  'synopsis': USAGE}] + TRANSFER_ARGS
+                  'synopsis': USAGE}] + TRANSFER_ARGS + [METADATA]
```
Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: Probably would want to add ``METADATA_DIRECTIVE`` as well.

File: awscli/customizations/s3/utils.py [hunk 5]
```diff
@@ -638,6 +641,13 @@ def _permission_to_param(cls, permission):
                          'read|readacl|writeacl|full')
 
     @classmethod
+    def _set_metadata_params(cls, request_params, cli_params):
+        if cli_params.get('metadata'):
+            request_params['Metadata'] = cli_params['metadata']
+            if not cli_params.get('metadata-directive'):
```
Comment (kyleknap) [inline awscli/customizations/s3/utils.py]: Will autofilling ``MetadataDirective`` work for ``CreateMultipartUpload`` and ``PutObject``? It does not look like it. May have to add edge cases for that.

File: tests/integration/customizations/s3/test_plugin.py [hunk 6]
```diff
@@ -439,6 +439,20 @@ def test_copy_large_file_signature_v4(self):
...
+        bucket_name = self.create_bucket()
+        key = 'foo.txt'
+        filename = self.files.create_file(key, contents='')
+        p = aws('s3 cp %s s3://%s/%s --metadata keyname=value' %
+                (filename, bucket_name, key))
+        self.assert_no_errors(p)
+        response = self.head_object(bucket_name, key)
+        # These values should have the metadata of the source object
+        self.assertEqual(response['Metadata'].get('keyname'), 'value')
+
```
Comment (kyleknap) [inline tests/integration/customizations/s3/test_plugin.py]: I would really like to see test cases (functional or integration-wise) that test supplying metadata for uploads and multipart transfers. It would answer my previous questions about metadata-directive.

**Comment by kyleknap** [comment]
Almost there. There are still a couple edge cases that I do not think you handle. For these type of edgecases it never hurts to add an integration or functional test to make sure it works through and through.

File: awscli/customizations/s3/utils.py [hunk 7]
```diff
@@ -638,6 +642,15 @@ def _permission_to_param(cls, permission):
                          'read|readacl|writeacl|full')
 
     @classmethod
+    def _set_metadata_params(cls, request_params, cli_params):
+        if cli_params.get('metadata'):
+            request_params['Metadata'] = cli_params['metadata']
+
+    @classmethod
+    def _auto_populate_metadata_directive(cls, request_params):
+        request_params['MetadataDirective'] = 'REPLACE'
```
Comment (kyleknap) [inline awscli/customizations/s3/utils.py]: We only want to replace`` MetadataDirective`` if it is not currently in ``request_params`` correct?

Comment (kyleknap) [inline awscli/customizations/s3/utils.py]: Also only autopopulate it if ``Metadata`` is present in ``request_params``. I think that is why the tests are failing.

**Comment by kyleknap** [comment]
I think once you fix the comment I had. It should be good to ship.

**Comment by JordonPhillips** [comment]
fixed

**Comment by kyleknap** [comment]
Thanks! Looks good :ship:
